BABEL-1199 Support DATABASE_PRINCIPAL_ID() T-SQL function

pratikzode opened the pull request:
Now it supports
SELECT DATABASE_PRINCIPAL_ID();
Retrieve the ID of the current user.

SELECT DATABASE_PRINCIPAL_ID(‘principal_name’);
Retrieve the ID of the specific database principal.

Task: BABEL-1199

pratikzode committed:
BABEL-1199 Support DATABASE_PRINCIPAL_ID() T-SQL function

SELECT DATABASE_PRINCIPAL_ID();
Retrieve the ID of the current user.

SELECT DATABASE_PRINCIPAL_ID(‘principal_name’);
Retrieve the ID of the specific database principal.

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Modify test cases

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Modify test cases

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Modify expected_dependency.out

Removed -Function sys.user_name(oid) from the file

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

xhfanhe commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -1707,6 +1707,20 @@ LANGUAGE SQL IMMUTABLE STRICT PARALLEL SAFE;
 CREATE OR REPLACE FUNCTION sys.rowcount()
 RETURNS INT AS 'babelfishpg_tsql' LANGUAGE C STABLE;
 
+CREATE OR REPLACE FUNCTION Database_principal_id(user_name TEXT)
```
Do we want to specify sys schema?

xhfanhe replied:
Also, input type should be sysname

pratikzode replied:
yes have changed it

xhfanhe commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -1707,6 +1707,20 @@ LANGUAGE SQL IMMUTABLE STRICT PARALLEL SAFE;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE FUNCTION Database_principal_id()
```
same as above

macwoody commented on test/JDBC/input/BABEL-1199-vu-verify.sql:
```diff
@@ -0,0 +1,19 @@
+
+
+SELECT user_name(DATABASE_PRINCIPAL_ID('db_owner'));
+SELECT user_name(DATABASE_PRINCIPAL_ID('testuser'));
```
please also test dbo and guest users

pratikzode replied:
Added it

macwoody commented on test/JDBC/input/BABEL-1199-vu-prepare.sql:
```diff
@@ -0,0 +1,29 @@
+
+CREATE LOGIN testuser WITH PASSWORD = 'testpassword';
+CREATE USER testuser FOR LOGIN testuser;
+GO
```
please also create a role

macwoody commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql:
```diff
@@ -842,6 +842,20 @@ CREATE OR REPLACE FUNCTION sys.radians(IN arg1 NUMERIC)
+CREATE OR REPLACE FUNCTION Database_principal_id(user_name TEXT)
+RETURNS OID AS $$
+BEGIN
+    RETURN sys.user_id(LOWER(user_name));
```
from my POV, database_principal_id is equivalent to user_id. 
Could you please point out what is the additional behavior you would like to add on top of it? 
Isn't case sensitivity handled internally by user_id implementation?

pratikzode replied:
Yes removed the LOWER

lejaokri commented on test/JDBC/input/BABEL-1199-vu-prepare.sql:
```diff
@@ -1,25 +1,29 @@
 
 CREATE VIEW dbo.view_db_owner_principal_id AS
-SELECT DATABASE_PRINCIPAL_ID('db_owner');
+SELECT user_name(DATABASE_PRINCIPAL_ID('db_owner'));
```
is there already a test with invalid user?

pratikzode committed:
BABEL-1199 Modify function file and test cases

Have added input type as sysname instead of text.
Have modified the RETURN sys.user_id(LOWER(user_name)); as RETURN sys.user_id(user_name); as user_id handles case sensitivity internally.
Also tested for test dbo and guest users.
Have created a role

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Modify the function

Combined two args in one

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

surendravishnoi commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -386,6 +386,7 @@ AVG-Aggregate-Dep
 bbf_view_def
 BABEL_OBJECT_ID
 BABEL-3802
+BABEL-1199
```
this will not trigger 3.0 to 3.1 upgrade script. Please add it to a schedule which will trigger 3.0 to 3.1 upgrade as well.

pratikzode committed:
BABEL-1199 Modify the function

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql:
```diff
@@ -842,6 +842,11 @@ CREATE OR REPLACE FUNCTION sys.radians(IN arg1 NUMERIC)
 RETURNS numeric  AS 'babelfishpg_tsql','numeric_radians' LANGUAGE C STRICT IMMUTABLE PARALLEL SAFE;
 GRANT EXECUTE ON FUNCTION sys.radians(NUMERIC) TO PUBLIC;
 
+CREATE OR REPLACE FUNCTION Database_principal_id(IN user_name TEXT DEFAULT NULL)
```
Input type is sysname
and we need upgrade scriopt of 3.1->3.2 for this, but please hold on the change , since we don't have it now

pratikzode committed:
BABEL-1199 Changed the role name in test case

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Changed the Input type to sysname

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>
